wit/atomic.wit

```diff
@@ -11,12 +11,36 @@
 interface atomics {
   	use store.{bucket, error};
 
+	/// The error returned by a CAS operation
+	variant cas-error {
+		/// A store error occurred when performing the operation
+		store-error(error),
+		/// The CAS operation failed because the value was too old. This returns a new CAS handle
+		/// for easy retries. Implementors MUST return a CAS handle that has been updated to the
+		/// latest version or transaction.
+		cas-failed(cas),
+	}
+
+	/// A handle to a CAS (compare-and-swap) operation.
+	resource cas {
+		/// Construct a new CAS operation. Implementors can map the underlying functionality
+		/// (transactions, versions, etc) as desired.
+		new: static func(bucket: borrow<bucket>, key: string) -> result<cas, error>;
+		/// Get the current value of the key (if it exists). This allows for avoiding reads if all
+		/// that is needed to ensure the atomicity of the operation
+		current: func() -> result<option<list<u8>>, error>;
+	}
```

Comment:
I am curious to hear your motivations / rational behind the cas resource design? My feeling for that when I first saw it was complicated and more unintuitive than a simple operation like `compare-and-swap: func(key: string, old: option<list<u8>>, new: option<list<u8>>) -> result<bool, error>;`

Reply:
This gives a lot more flexibility for how implementors can solve this. As seen in the [survey of keyvalue stores I did](https://github.com/WebAssembly/wasi-keyvalue/issues/44#issuecomment-2257037553), the most common things were transactions or revision_id-based options, along with some direct comparing and swapping. Representing this as a resource allows implementations to keep track of state for the operation. So for each of the 3 aforementioned types of operation, this is what is would roughly look like:

- Transactions: An implementation can start a transaction and link it to a resource. You can still fetch the current value if you wish to do a comparison, but otherwise can use the guarantee of the transaction to ensure that you aren't clobbering a write
- Revision IDs: Implementations can fetch the current revision of the key along with the value (or lazily load the value if `current` is called) and keep that revision ID around as its state. Then when the value is swapped it can use the revision ID
- Direct compare and swap: The implementation wouldn't actually have any state but could fetch the value and do the comparison on the back end. As mentioned, this was by far the least used option among all the stores I surveyed

Forcing everyone to do a direct compare and swap means they have to a) fetch the value and b) send the other value. This could be much less efficient, especially if the values are large. It basically forces everything to implement it using a much less efficient way (that is much more vulnerable to the ABA problem you mentioned below) even though it is the least common option. The other options at least reduce the possibility of or eliminate the ABA problem as well, addressing your other comment)